src/irmin-pack/unix/dispatcher.ml

```diff
@@ -43,25 +43,23 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
   (** [mapping] is a map from global offset to (offset,len) pairs in the prefix
       file *)
 
-  let load_mapping io =
-    let open Result_syntax in
-    let open Int63 in
-    let open Int63.Syntax in
+  let load_mapping path =
     let mapping = ref Intmap.empty in
-    let poff = ref zero in
+    let poff = ref Int63.zero in
     let f ~off ~len =
       mapping := Intmap.add off { poff = !poff; len } !mapping;
-      poff := !poff + of_int len
+      poff := Int63.(Syntax.(!poff + of_int len))
     in
-    let* () = Mapping_file.iter io f in
+    let arr = Mapping_file.load_mapping_as_mmap path in
```

Comment:
Is it that both `arr` and the `mapping` are in memory, now?

Reply:
The arr is mmap'ed, An mmap is handled by the OS, and doesn't need to hold all the data in memory at once. For example, if I have 1GB of memory, and I mmap a 4GB file, the entire file isn't kept in memory: the OS just makes it look that way. In this code, it looks like arr and mapping are both "live" (and so potentially we have a lot of memory overhead) but this isn't what is happening. If arr was a normal array, the situation would be different! Another thing: I think we are moving to dropping the mapping (Map.t) and just performing binary search over the arr. So only the arr will exist, which is much more compact than the equivalent Map.t (3x - 6x smaller maybe). (Please correct me if any of this seems wrong.)